src/commands.jl

```diff
@@ -0,0 +1,183 @@
+when it hits a `return` statement) or a reference to a breakpoint.
+In the latter case, `leaf(frame)` returns the frame in which it hit the breakpoint.
+
+`recurse` controls call evaluation; `recurse = Compiled()` evaluates :call expressions
+by normal dispatch, whereas the default `recurse = finish_and_return!` uses recursive interpretation.
+"""
+function finish!(@nospecialize(recurse), frame::Frame, istoplevel::Bool=false)
+    local pc
+    while true
+        pc = step_expr!(recurse, frame, istoplevel)
+        (pc == nothing || isa(pc, BreakpointRef)) && return pc
+        shouldbreak(frame, pc) && return BreakpointRef(frame.framecode, pc)
+    end
+end
+finish!(frame::Frame, istoplevel::Bool=false) = finish!(finish_and_return!, frame, istoplevel)
+
+"""
+    ret = finish_and_return!(recurse, frame, istoplevel::Bool=false)
+    ret = finish_and_return!(frame, istoplevel::Bool=false)
+
+Call [`JuliaInterpreter.finish!`](@ref) and pass back the return value. If execution
+pauses at a breakpoint, the reference to the breakpoint is returned.
+"""
+function finish_and_return!(@nospecialize(recurse), frame::Frame, istoplevel::Bool=false)
+    pc = finish!(recurse, frame, istoplevel)
+    isa(pc, BreakpointRef) && return pc
+    return get_return(frame)
+end
+finish_and_return!(frame::Frame, istoplevel::Bool=false) = finish_and_return!(finish_and_return!, frame, istoplevel)
+
+"""
+    bpref = dummy_breakpoint(recurse, frame::Frame)
+
+Return a fake breakpoint. This can be useful as the `recurse` argument to `evaluate_call!`
+(or any of the higher-order commands) to ensure that you return immediately after stepping
+into a call.
+"""
+dummy_breakpoint(@nospecialize(recurse), frame::Frame) = BreakpointRef(frame.framecode, 0)
+
+"""
+    ret = finish_stack!(recurse, frame, istoplevel=false)
+    ret = finish_stack!(frame, istoplevel=false)
+
+Unwind the callees of `frame`, finishing each before returning to the caller.
+`frame` itself is also finished
+If execution hits a breakpoint, `ret` will be a reference to the breakpoint.
+"""
+function finish_stack!(@nospecialize(recurse), frame::Frame, istoplevel::Bool=false)
+    frame0 = frame
+    frame = leaf(frame)
+    while true
+        ret = finish_and_return!(recurse, frame, istoplevel)
+        isa(ret, BreakpointRef) && return ret
+        frame === frame0 && return ret
+        recycle(frame)
+        frame = caller(frame)
+        frame === nothing && return ret
+        pc = frame.pc
+        if isassign(frame, pc)
+            lhs = getlhs(pc)
+            do_assignment!(frame, lhs, ret)
+        end
+        pc += 1
+        frame.pc = pc
+        shouldbreak(frame) && return BreakpointRef(frame.framecode, pc)
+    end
+    return frame
+end
+finish_stack!(frame::Frame, istoplevel::Bool=false) = finish_stack!(finish_and_return!, frame, istoplevel)
+
+"""
+    next_until!(predicate, recurse, frame, istoplevel=false)
```

Comment:
Is it worth saying that `pc` is returned here (so you know e.g. that this can return a breakpoint).